app/tests/profiles_tests/test_models.py

```diff
@@ -69,3 +81,80 @@ def test_submit_newsletter_preference(client):
         user=u2,
     )
     assert response.status_code == 403
+
+
+@pytest.mark.django_db
+def test_dispatch_notifications_email():
+    user = UserFactory()
+    site = Site.objects.get()
+    assert not user.user_profile.notification_email_last_sent_at
+
+    user.user_profile.dispatch_unread_notifications_email(
+        site=site, unread_notification_count=1
+    )
+    assert user.user_profile.notification_email_last_sent_at
+    assert len(mail.outbox) == 1
+    assert mail.outbox[0].to == [user.email]
+
+
+@pytest.mark.parametrize(
+    "subscription_type,subscription_attr,subscription_preference,unsubscribe_viewname,expectation",
+    (
+        (
+            EmailSubscriptionTypes.NOTIFICATION,
+            "notification_email_choice",
+            NotificationEmailOptions.DAILY_SUMMARY,
+            "notification-unsubscribe",
+            nullcontext(),
+        ),
+        (
+            EmailSubscriptionTypes.NOTIFICATION,
+            "notification_email_choice",
+            NotificationEmailOptions.INSTANT,
+            "notification-unsubscribe",
+            nullcontext(),
+        ),
+        (
+            EmailSubscriptionTypes.NOTIFICATION,
+            "notification_email_choice",
+            NotificationEmailOptions.DISABLED,
+            "notification-unsubscribe",
+            pytest.raises(ValueError),
+        ),
+        (
+            EmailSubscriptionTypes.NEWSLETTER,
+            "receive_newsletter",
+            True,
+            "newsletter-unsubscribe",
+            nullcontext(),
+        ),
+        (
+            EmailSubscriptionTypes.NEWSLETTER,
+            "receive_newsletter",
+            False,
+            "newsletter-unsubscribe",
+            pytest.raises(ValueError),
+        ),
+        (EmailSubscriptionTypes.SYSTEM, None, None, None, nullcontext()),
+    ),
+)
+@pytest.mark.django_db
+def test_get_unsubscribe_link(
+    subscription_type,
+    subscription_attr,
+    subscription_preference,
+    unsubscribe_viewname,
+    expectation,
+):
+    user = UserFactory()
+    if subscription_preference:
+        setattr(user.user_profile, subscription_attr, subscription_preference)
+    with expectation:
+        link = user.user_profile.get_unsubscribe_link(
+            subscription_type=subscription_type
+        )
+        if link:
+            assert link == reverse(
+                unsubscribe_viewname,
+                kwargs={"token": user.user_profile.unsubscribe_token},
+            )
```

Comment:
Only 1 action should be performed in a `with pytest.raises` block. Each statement should be in its own context. That way, the test ensures that the exception is raised only in the exact statement where you expect it. Remember that the exception is raised and caught, so any code after the exception will not be executed.